services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/gdpr/KortistoGdprServiceITest.kt

```diff
@@ -200,20 +207,122 @@ class KortistoGdprServiceITest(
+                rakennuttaja { addYhteyshenkilo(it, kayttaja) }
+            }
+
+            val result = gdprService.canDelete(USERID)
+
+            assertThat(result).isTrue()
+        }
+
+        @Test
+        fun `throws exception when there's an active application`() {
+            val hanke = hankeFactory.builder(USERID).withHankealue().saveEntity()
+            val kayttaja = hankeKayttajaService.getKayttajaByUserId(hanke.id, USERID)!!
+            hakemusFactory.builder(USERID, hanke).saveWithYhteystiedot {
+                asianhoitaja { addYhteyshenkilo(it, kayttaja) }
+            }
+            val activeHakemus =
+                hakemusFactory.builder(USERID, hanke).inHandling().saveWithYhteystiedot {
+                    rakennuttaja { addYhteyshenkilo(it, kayttaja) }
+                }
+
+            val failure = assertFailure { gdprService.canDelete(USERID) }
+
+            failure.given { e ->
+                val messages = (e as DeleteForbiddenException).errors.map { it.message.fi }
+                assertThat(messages).single().contains(activeHakemus.applicationIdentifier!!)
+            }
+        }
+
+        @Test
+        fun `returns true when the user is not an admin and not on any active applications`() {
+            val adminUserId = "admin"
+            val hanke = hankeFactory.builder(adminUserId).withHankealue().saveEntity()
+            val adminKayttaja = hankeKayttajaService.getKayttajaByUserId(hanke.id, adminUserId)!!
+            val targetKayttaja =
+                hankekayttajaFactory.saveIdentifiedUser(
+                    hanke.id,
+                    userId = USERID,
+                    kayttooikeustaso = Kayttooikeustaso.HANKEMUOKKAUS
+                )
+            hakemusFactory.builder(adminUserId, hanke).saveWithYhteystiedot {
+                asianhoitaja { addYhteyshenkilo(it, targetKayttaja) }
+            }
+            hakemusFactory.builder(adminUserId, hanke).inHandling().saveWithYhteystiedot {
+                rakennuttaja { addYhteyshenkilo(it, adminKayttaja) }
+            }
+
+            val result = gdprService.canDelete(USERID)
+
+            assertThat(result).isTrue()
+        }
+
+        @Test
+        fun `throws exception when user is the only admin and there's an active application on the hanke`() {
+            val hanke = hankeFactory.builder(USERID).withHankealue().saveEntity()
+            hakemusFactory.builder(USERID, hanke).saveWithYhteystiedot { asianhoitaja() }
+            val activeHakemus =
+                hakemusFactory.builder(USERID, hanke).inHandling().saveWithYhteystiedot {
+                    rakennuttaja()
+                }
+
+            val failure = assertFailure { gdprService.canDelete(USERID) }
+
+            failure.given { e ->
+                val messages = (e as DeleteForbiddenException).errors.map { it.message.fi }
+                assertThat(messages).single().contains(activeHakemus.applicationIdentifier!!)
```

Comment:
Maybe this test should assert that there is something different in the result compared to test `throws exception when there's an active application`? Or is there any difference between the results in these two tests?

Reply:
The errors they return are exactly the same. So a request might fail on both for the same hakemus. The response is mapped to a set to get rid of the duplicates.